### PR TITLE
Improve LIKE performance

### PR DESCRIPTION
**Pull request by martint**
Implements a dedicated NFA/DFA-based matcher for LIKE expressions.

Before:
```
  (pattern)  Mode  Cnt    Score    Error  Units
          %  avgt   90  249.201 ±  1.761  ns/op
         _%  avgt   90  290.866 ±  0.947  ns/op
         %_  avgt   90  299.377 ±  2.079  ns/op
       abc%  avgt   90  244.724 ±  2.004  ns/op
       %abc  avgt   90  139.165 ±  7.407  ns/op
      _____  avgt   90   25.508 ±  0.161  ns/op
abc%def%ghi  avgt   90  209.570 ± 11.658  ns/op
  %abc%def%  avgt   90  447.848 ±  0.739  ns/op
```

After:
```
          %  avgt   90    3.989 ±  0.012  ns/op
         _%  avgt   90    3.969 ±  0.014  ns/op
         %_  avgt   90    3.958 ±  0.013  ns/op
       abc%  avgt   90    5.501 ±  0.019  ns/op
       %abc  avgt   90    5.861 ±  0.015  ns/op
      _____  avgt   90    7.096 ±  0.028  ns/op
abc%def%ghi  avgt   90   84.293 ±  0.623  ns/op
  %abc%def%  avgt   90  152.031 ±  0.247  ns/op
```

## Documentation

(x) No documentation is needed.

## Release notes

(x) Release notes entries required with the following suggested text:

```markdown
# General
* Improve performance of `LIKE` expressions. ({issue}`13479`)
```

### CODE REVIEW

File: core/trino-main/src/main/java/io/trino/likematcher/DenseDfaMatcher.java
Comment (martint) [reply core/trino-main/src/main/java/io/trino/likematcher/DenseDfaMatcher.java:37]: No, that's an oversight.

The only cases where we should do that is when we do an RNN and immediately perform an operation that might fails with NPE. In that case, the JVM will produce better error messages that the RNN check.

File: core/trino-main/src/main/java/io/trino/likematcher/DFA.java
Comment (martint) [inline core/trino-main/src/main/java/io/trino/likematcher/DFA.java:27]: Still deciding whether this should be a Record or a Class. The issue is that for a Record, the `states` and `transitions` will reference the provided instances w/o an intervening defensive copy or immutability guarantees.

Comment (martint) [reply core/trino-main/src/main/java/io/trino/likematcher/DFA.java:27]: I'm concerned that it requires that extra code in construction, which seems a bit unnatural.

Comment (findepi) [reply core/trino-main/src/main/java/io/trino/likematcher/DFA.java:27]: Making it a Class doesn't let you get away from that.
Alternatively you could expect user to provide `ImmutableList` and `ImmutableMap` explicitly.
I wouldn't mind personally, but it would be a deviation from what we have practiced so far.

File: core/trino-main/src/main/java/io/trino/likematcher/LikeMatcher.java
Comment (phd3) [inline core/trino-main/src/main/java/io/trino/likematcher/LikeMatcher.java:148]: Can we add a comment here as well ? multiplying by 256 on RHS isn't obvious atleast to me

File: core/trino-main/src/main/java/io/trino/likematcher/DenseDfaMatcher.java
Comment (phd3) [inline core/trino-main/src/main/java/io/trino/likematcher/DenseDfaMatcher.java:23]: :)

File: core/trino-main/src/main/java/io/trino/likematcher/LikeMatcher.java
Comment (phd3) [inline core/trino-main/src/main/java/io/trino/likematcher/LikeMatcher.java:226]: a bit tangential, does sql spec allow % and _ as escapes too?

File: core/trino-main/src/main/java/io/trino/likematcher/NFA.java
Comment (martint) [reply core/trino-main/src/main/java/io/trino/likematcher/NFA.java:31]: It's the parameter. However, for records, there's an implicit `this.states = states` performed at the end of the constructor.

File: core/trino-main/src/main/java/io/trino/likematcher/NFA.java
Comment (martint) [reply core/trino-main/src/main/java/io/trino/likematcher/NFA.java:32]: No, just leftover from when I was experimenting outside of the project and didn't have Guava.

File: core/trino-main/src/main/java/io/trino/likematcher/LikeMatcher.java
Comment (martint) [reply core/trino-main/src/main/java/io/trino/likematcher/LikeMatcher.java:52]: I omitted them because it's a private constructor and all the necessary checks are done in the single static method that calls it.

File: core/trino-main/src/main/java/io/trino/likematcher/NFA.java
Comment (martint) [reply core/trino-main/src/main/java/io/trino/likematcher/NFA.java:30]: No, I'll add them

File: core/trino-main/src/test/java/io/trino/sql/planner/TestLiteralEncoder.java
Comment (martint) [reply core/trino-main/src/test/java/io/trino/sql/planner/TestLiteralEncoder.java:242]: They are used to make sure we can encode patterns properly when converting expressions to textual form. It's because like and regex patterns get converted to specialized types under the covers, so reversing that takes a bit of work.

File: core/trino-main/src/main/java/io/trino/likematcher/LikeMatcher.java
Comment (martint) [reply core/trino-main/src/main/java/io/trino/likematcher/LikeMatcher.java:143]: It's not significant in general that it's worth tracking. The memory utilized will be proportional to the length of the pattern * 256, so it will be just a few KB at most in general.

File: core/trino-main/src/main/java/io/trino/likematcher/LikeMatcher.java
Comment (martint) [reply core/trino-main/src/main/java/io/trino/likematcher/LikeMatcher.java:193]: No, but after re-reading it, I realize it's not very clear. I'll just rename the method to make the intention clear and remove the comment.

File: core/trino-main/src/main/java/io/trino/likematcher/LikeMatcher.java
Comment (martint) [reply core/trino-main/src/main/java/io/trino/likematcher/LikeMatcher.java:226]: Yes, it allows any character as an escape.

File: core/trino-main/src/main/java/io/trino/likematcher/LikeMatcher.java
Comment (martint) [reply core/trino-main/src/main/java/io/trino/likematcher/LikeMatcher.java:148]: This is explained in DenseDfaMatcher, which this class is constructing

**Review by jirassimok — COMMENTED** [review]
Looking through how this PR works, I think there's something of an extra step here in the construction of the DFA. It looks like the `DFA` class is only used as a transitory state between `NFA` and `DenseDfaMatcher`, the latter of which actually _functions_ as a DFA.

I wonder if replacing the current `DFA` with a DFA builder and renaming `DenseDfaMatcher` to `DFA` might be a better idea.

Either way, I think the part of `LikeMatcher.compile` that essentially compiles a DFA should be moved to its own method on the current `DFA` class.

**Comment by martint** [comment]
> Looking through how this PR works, I think there's something of an extra step here in the construction of the DFA. It looks like the DFA class is only used as a transitory state between NFA and DenseDfaMatcher, the latter of which actually functions as a DFA.

The DFA class is an abstract description of the DFA (useful for understanding the states and transitions). The one used by DfaMatcher is a physical version designed for efficient matching.

> Either way, I think the part of LikeMatcher.compile that essentially compiles a DFA should be moved to its own method on the current DFA class.

You make a good point. That bit of code should actually go in DenseDfaMatcher, which should provide a method for constructing a instance of that class from the abstract representation of the DFA.

File: core/trino-main/src/main/java/io/trino/likematcher/LikeMatcher.java
Comment (martint) [reply core/trino-main/src/main/java/io/trino/likematcher/LikeMatcher.java:101]: It's not that simple. A bounded any needs to match UTF-8 characters, which can vary in length between 1 and 4 bytes and needs to through a DFA.

I have a few other optimizations in mind (and some prototypes) that involve switching between DFA and exact literal matches at the right places by splitting the pattern into sections, but it gets very complicated very quickly and make it hard to get the best raw performance for smaller patterns.

**Review by phd3 — COMMENTED** [review]
still reviewing

File: core/trino-main/src/main/java/io/trino/type/LikeFunctions.java
Comment (martint) [reply core/trino-main/src/main/java/io/trino/type/LikeFunctions.java:73]: Hmm.. good catch. It's probably a logical merge conflict from when I rebased my old branch onto latest master.

File: core/trino-main/src/main/java/io/trino/type/LikePatternType.java
Comment (martint) [reply core/trino-main/src/main/java/io/trino/type/LikePatternType.java:64]: It should work. The pattern gets written to a block as `string | int?` for the pattern and the optional escape. But maybe you're seeing something I'm not?

In any case, LikePatternType is a synthetic type and this method is only used to serialize / deserialize patterns to ship them to workers.

File: core/trino-main/src/test/java/io/trino/operator/scalar/BenchmarkLike.java
Comment (martint) [reply core/trino-main/src/test/java/io/trino/operator/scalar/BenchmarkLike.java:191]: No, that's correct. The string is a format string (per String.format), and `%` needs to be escaped.

File: core/trino-main/src/main/java/io/trino/type/LikePatternType.java
Comment (phd3) [reply core/trino-main/src/main/java/io/trino/type/LikePatternType.java:64]: `getObject(position)` doesn’t us you the LikePattern  at position without caller side manipulation. Also, we add 2 or 3 elements respectively in case of  with/without escape. (appendNull.closeEntry adds two vs writeInt.closeEntry adds one) so position manipulation may not be straightforward. I see the point that this is all synthetic/internal, but curious if the usage for more than one elements is kosher

Comment (phd3) [reply core/trino-main/src/main/java/io/trino/type/LikePatternType.java:64]:
```
        BlockBuilder blockBuilder = LIKE_PATTERN.createBlockBuilder(new PageBuilderStatus().createBlockBuilderStatus(), 10);
        LIKE_PATTERN.writeObject(blockBuilder, LikeMatcher.compile("abcd", Optional.of('g')));
        LIKE_PATTERN.writeObject(blockBuilder, LikeMatcher.compile("abc"));
        Block block = blockBuilder.build(); -- contains 5 positions (2 + 3) 
        //  incorrect mapping to previous entry's escape. 
        LIKE_PATTERN.getObject(block, 1);
```

File: core/trino-main/src/main/java/io/trino/likematcher/LikeMatcher.java
Comment (phd3) [inline core/trino-main/src/main/java/io/trino/likematcher/LikeMatcher.java:306]: nit: I think this works because `any(0, false)` is not allowed by definition because it's just an empty string and not a transition. would leave it to you if worth  enforcing that in Any constructor

**Review by phd3 — APPROVED** [review]
lgtm, just one pending comment about likepattern serde

File: core/trino-main/src/main/java/io/trino/type/LikePatternType.java
Comment (martint) [reply core/trino-main/src/main/java/io/trino/type/LikePatternType.java:64]: Oh, yeah, that's completely broken. I actually meant it to write at different offsets but somehow I ended using positions instead. I'll fix it.

File: core/trino-main/src/main/java/io/trino/likematcher/LikeMatcher.java
Comment (martint) [reply core/trino-main/src/main/java/io/trino/likematcher/LikeMatcher.java:324]: Yes, that's a good point. I added a check to the Any constructor and adjusted code elsewhere that could result in `any(0, false)` after pattern optimization.

**Review by phd3 — APPROVED** [review]

**Pull request merged**
